### DISCUSSION TITLE
How to print human readable time from a variable which has a timestamp in ns?

### DISCUSSION THREAD

**Discussion by rmilkowski** [bpftrace/bpftrace#2486] [Q&A]
Hi,

strftime(..., t1) assumes t1 is nanoseconds since boot, while in my case it is a timestamp in ns.

Also, how do I get current timestamp in ns (nsec is since boot)?

**Comment by viktormalik** [comment]
You can use the `nsecs` builtin.

**Comment by rmilkowski** [comment]
No, as nsecs is since boot. I found a way to do it but it is cumbersome.

**Comment by viktormalik** [comment]
Ah, right. If you just want the current timestamp, you can use the `time()` call.

**Comment by rmilkowski** [comment]
No, as it only returns a string and has 1s resolution. 
I’ll post later how I ended up doing it.

**Reply by sanmai-NL** [reply]
@rmilkowski Do you have an update?

**Reply by rmilkowski** [reply]
Yes, see http://milek.blogspot.com/2023/03/bpf-unix-timestamp.html

**Reply by viktormalik** [reply]
There is now the [bpf_ktime_get_tai_ns helper](https://elixir.bootlin.com/linux/v6.2/source/kernel/bpf/helpers.c#L204) that allows BPF programs to access the TAI clock. Would that be sufficient for your needs?

I opened #2512 to add support for that into bpftrace but didn't have time to implement it so far.
